src/cow-shed/CowShedSdk.ts

```diff
@@ -0,0 +1,157 @@
+  factoryOptions?: ICoWShedOptions
+}
+
+export class CowShedSdk {
+  static COW_SHED_CACHE = new Map<SupportedChainId, CowShedHooks>()
+
+  constructor(private options: CowShedSdkOptions = {}) {}
+
+  getCowShedAccount(chainId: SupportedChainId, ownerAddress: string): string {
+    const cowShedHooks = CowShedSdk.getCowShedHooks(chainId, this.options?.factoryOptions)
+    return cowShedHooks.proxyOf(ownerAddress)
+  }
+
+  /**
+   * Encodes multiple calls into a single pre-authorized call to the cow-shed factory.
+   *
+   * This single call will create the cow-shed account if it doesn't exist yet, then will execute the calls.
+   *
+   * @returns pre-authorized multicall details
+   */
+  async signCalls({
+    calls,
+    signer: signerParam,
+    chainId,
+    nonce = CowShedSdk.getNonce(),
+    deadline = NON_EXPIRING_DEADLINE,
+    defaultGasLimit,
+    signingScheme = SigningScheme.EIP712,
+  }: SignAndEncodeTxArgs): Promise<CowShedCall> {
+    // Get the cow-shed for the wallet
+    const cowShedHooks = CowShedSdk.getCowShedHooks(chainId)
+
+    // Get the signer, or throw an error if not provided
+    const signerLike = signerParam || this.options?.signer
+    if (!signerLike) {
+      throw new Error('Signer is required')
+    }
+    const signer = getSigner(signerLike)
+    const ownerAddress = await signer.getAddress()
+
+    const cowShedAccount = cowShedHooks.proxyOf(ownerAddress)
+
+    // Sign the calls using cow-shed's owner
+    const signature = await cowShedHooks.signCalls(calls, nonce, deadline, signer, signingScheme)
+
+    //  Get the signed transaction's calldata
+    const callData = cowShedHooks.encodeExecuteHooksForFactory(calls, nonce, deadline, ownerAddress, signature)
+
+    // Estimate the gas limit for the transaction
+    const cowShedFactoryAddress = cowShedHooks.getFactoryAddress()
+    const factoryCall = {
+      to: cowShedFactoryAddress,
+      data: callData,
+      value: BigInt(0),
+    }
+    const gasEstimate = await signer.estimateGas(factoryCall).catch((error) => {
+      console.error('Error estimating gas for the cow-shed call: ' + JSON.stringify(factoryCall), error)
+
+      // Return the default gas limit if provided
+      if (defaultGasLimit) {
+        return defaultGasLimit
+      }
+
+      // Re-throw the error if no default gas limit is provided
+      throw error
+    })
+
+    // Return the details, including the signed transaction data
+    return {
+      cowShedAccount,
+      signedMulticall: factoryCall,
+      gasLimit: gasEstimate.toBigInt(),
+    }
+  }
+
+  protected static getCowShedHooks(chainId: SupportedChainId, customOptions?: ICoWShedOptions) {
+    let cowShedHooks = CowShedSdk.COW_SHED_CACHE.get(chainId)
+
+    if (cowShedHooks) {
+      // Return cached cow-shed hooks
+      return cowShedHooks
+    }
+
+    // Create new cow-shed hooks and cache it
+    cowShedHooks = new CowShedHooks(chainId, customOptions)
+    CowShedSdk.COW_SHED_CACHE.set(chainId, cowShedHooks)
```

Comment:
Here we can face cache inconsistancy.
They only key of the cache is `chainId`, but there are also `customOptions`.
For example:
1. You get an instance like: `const hooks1 = CowShedSdk.getCowShedHooks(1, {factoryAddress: '0x...1'})`
2. You get another: `const hooks2 = CowShedSdk.getCowShedHooks(1, {factoryAddress: '0x...2'})`

In the `hooks2` you will actually get an instance with `factoryAddress='0x...1'`

Reply:
Sorry, I think the issue is still there. `customOptions` is still not a part of the cache key